aiohttp/connector.py

```diff
@@ -899,82 +907,93 @@ async def _resolve_host(
+        self._throttle_dns_futures[key] = futures = set()
+        if traces:
+            for trace in traces:
+                await trace.send_dns_cache_miss(host)
+        try:
             if traces:
                 for trace in traces:
-                    await trace.send_dns_cache_miss(host)
+                    await trace.send_dns_resolvehost_start(host)
+
+            coro = self._resolver.resolve(host, port, family=self._family)
+            #
+            # If there is no resolution in progress, we need to start one.
+            #
+            # In this case we need to create a task to ensure that we can shield
+            # the task from cancellation as cancelling this lookup should not cancel
+            # the underlying lookup or else the cancel event will get broadcast to
+            # all the waiters across all connections.
+            #
+            loop = asyncio.get_running_loop()
+            if sys.version_info >= (3, 12):
+                # Optimization for Python 3.12, try to send immediately
+                resolved_host_task = asyncio.Task(coro, loop=loop, eager_start=True)
+            else:
+                resolved_host_task = loop.create_task(coro)
+
+            if not resolved_host_task.done():
+                self._resolve_host_tasks.add(resolved_host_task)
+                resolved_host_task.add_done_callback(self._resolve_host_tasks.discard)
+
             try:
-                if traces:
-                    for trace in traces:
-                        await trace.send_dns_resolvehost_start(host)
+                addrs = await asyncio.shield(resolved_host_task)
+            except asyncio.CancelledError:
 
-                addrs = await self._resolver.resolve(host, port, family=self._family)
-                if traces:
-                    for trace in traces:
-                        await trace.send_dns_resolvehost_end(host)
+                def drop_exception(fut: "asyncio.Future[List[ResolveResult]]") -> None:
+                    with suppress(Exception, asyncio.CancelledError):
+                        fut.result()
 
-                self._cached_hosts.add(key, addrs)
-                self._throttle_dns_events[key].set()
-            except BaseException as e:
-                # any DNS exception, independently of the implementation
-                # is set for the waiters to raise the same exception.
-                self._throttle_dns_events[key].set(exc=e)
+                resolved_host_task.add_done_callback(drop_exception)
                 raise
-            finally:
-                self._throttle_dns_events.pop(key)
 
-        return self._cached_hosts.next_addrs(key)
+            if traces:
+                for trace in traces:
+                    await trace.send_dns_resolvehost_end(host)
+
+            self._cached_hosts.add(key, addrs)  # type: ignore[possibly-undefined]
+            for fut in futures:
+                set_result(fut, None)
```

Comment:
We don't actually have any coverage for two requests in flight. This was hidden by the fact the previous event implementation had its own coverage.

Reply:
Considering how many different issues are in #9447 I think we should add a test for the following

- resolve while other resolve request in flight, in flight request is successful
- resolve while other resolve request in flight, in flight request is a failure
- resolve while other resolve request in flight, in flight request cancelled